### PR TITLE
Document Hasher spec decision from #42951

### LINKED ISSUE #42951: Spec request. Hasher: is write_u32 eqivalent to 4 calls of write_u8?
Currently `Hasher::write_u32` is implemented as `self.write(&unsafe { mem::transmute::<_, [u8; 4]>(i) })`. The question is: is it a requirement for all `Hasher` implementations?

In other words, is is true, that for each hasher implementation, call

```
hasher.write_32(0x01020304);
```

must be equivalent to

```
hasher.write_8(0x04);
hasher.write_8(0x03);
hasher.write_8(0x02);
hasher.write_8(0x01);
```

?

(assuming host is little-endian).

Why is it important.

Suppose, you want to implement very simple hasher with state updated like `state = state * 31 + update`.

If `Hasher` does not have requirement as above, hasher can be implemented by simply casting any operand to `u64` and avoiding dealing with unaligned/smaller than u64 integers:

```
impl Hasher for MyHasher {
    fn write_u8(&mut self, v: u8) {
        self.write_u64(v as u64);
    }

    fn write_u64(&mut self, v: u64) {
        self.state = (self.state * 31).wrapping_add(v);
    }
}
```

### PR DESCRIPTION
**Pull request by clarfonthey**
Since that ticket was closed without the decision actually being documented.

Fixes #42951.

### CODE REVIEW

File: library/core/src/hash/mod.rs [hunk 1]
```diff
@@ -240,6 +245,7 @@ pub use macros::Hash;
 /// [`finish`]: Hasher::finish
 /// [`write`]: Hasher::write
 /// [`write_u8`]: Hasher::write_u8
+/// [`write_u8`]: Hasher::write_u32
```
Comment (Amanieu) [inline library/core/src/hash/mod.rs:248]:
```suggestion
/// [`write_u32`]: Hasher::write_u32
```

**Review by Amanieu — COMMENTED** [review]
Could you also add a comment on `Hash::hash_slice` which says that `hash_slice` is not necessarily equivalent to multiple calls to `hash`.

File: library/core/src/hash/mod.rs (hunk 1 above)
Comment (clarfonthey) [reply library/core/src/hash/mod.rs:248]: 🤦🏻 thank you

**Comment by clarfonthey** [comment]
> Could you also add a comment on `Hash::hash_slice` which says that `hash_slice` is not necessarily equivalent to multiple calls to `hash`.

Since this doesn't really fit in the [`Hasher`] docs, I mentioned this on the method directly, including the `VecDeque` example. If anyone has any suggestions on how to better word this, feel free; I decided that more detail was better than less.

File: library/core/src/hash/mod.rs [hunk 2]
```diff
@@ -175,6 +175,21 @@ pub trait Hash {
 
     /// Feeds a slice of this type into the given [`Hasher`].
     ///
+    /// This method is meant as a convenience, but its implementation is
+    /// also explicitly left undefined. It isn't guaranteed to be
```
Comment (jyn514) [inline library/core/src/hash/mod.rs:179]:
```suggestion
    /// also explicitly left unspecified. It isn't guaranteed to be
```
Undefined behavior means "the compiler can delete large swaths of code", not "the implementation can return an arbitrary value". https://stackoverflow.com/a/2397995/7669110

File: library/core/src/hash/mod.rs [hunk 3]
```diff
@@ -175,6 +175,21 @@ pub trait Hash {
...
+    /// equivalent to repeated calls of [`hash`] and implementations of
+    /// [`Hash`] should keep that in mind and call [`hash`] themselves
+    /// if the slice isn't treated as a whole unit in the [`PartialEq`]
+    /// implementation.
+    ///
+    /// For example, a [`VecDeque`] implementation might naïvely call
+    /// [`as_slices`] and then [`hash_slice`] on each slice, but this
+    /// is wrong since the two slices can change with a call to
+    /// [`make_contiguous`] without affecting the [`PartialEq`]
+    /// implementation. Since these slices aren't treated as singular
```
Comment (Amanieu) [inline library/core/src/hash/mod.rs:189]:
```suggestion
    /// result. Since these slices aren't treated as singular
```

**Comment by clarfonthey** [comment]
Realised that I accidentally left this hanging and applied the suggested changes + rebased. Appears to be passing tests too but I'll wait for CI on that.

**Comment by Amanieu** [comment]
You need to point directly to the html file for doc links to std: `../../std/collections/struct.VecDeque.html`

**Pull request merged**